# miRDeep-P2_pipeline
This is a general pipeline for performing miRNA prediction and differential expression analysis using small RNA sequencing (sRNA-seq) data of plant samples. In this pipeline, the tool for miRNA prediction from sRNA-seq data is miRDeep-P2 (http://www.srnaworld.com/miRDP2/index.html). The output of miRDeep-P2 is parsed by parse_miRDeep_prediction.pl script in this pipeline. The parsed predicted miRNA sequences are used for  differential expression analysis. In addition, it is recommended to merge the predicted miRNA sequences with those present in the public domain, such as miRBase, for differential expression analysis, in order to increase the sensitivity. After getting the miRNA sequences, the sRNA-seq read data are mapped using Bowtie (http://bowtie-bio.sourceforge.net/index.shtml) and the read count for each miRNA sequence is generated using the bam2ref_counts.pl script in this pipeline. Finally, with the miRNA count data, DESeq2 (https://bioconductor.org/packages/release/bioc/html/DESeq2.html) is used to perform data filtering, normalization and differential expression analysis.

The details of using this pipeline are described as following.

## File
Clean sRNA-seq data in .fastq format for different samples, named `sample1.fq`, `sample2.fq` and etc.
Reference genome sequence in .fasta format, named `ref.fasta`
Index file of the reference genome seqeunce generated by Bowtie, e.g. using the following command

`bowtie-build ref.fasta ref.fasta`

## Tool


## 1. miRNA prediction using miRDeep-P2
The pipeline has been tested with miRDeep-P2 (v1.1.4), which is used for demonstration, and should be compatible with the other versions. Please refer to the tool's page for isntallation. To run miRDeep-P2, the following command can be used.

`miRDP2-v1.1.4_pipeline.bash -g ref.fasta -x ref.fasta -q -b fastq_list.txt -o result`

The file `fasta_list.txt` contains the absolute path to each of the .fastq file, with one line for one file. The output directory `result` contains the resulted results in its subdirectory for all samples, named sample1, sample2 and etc. Particularly, the files with suffix `_predictions`, e.g. `result/sample1/sample1_predictions`, are parsed to extract the predicted miRNA seqeunces and their corresponding precursor information. These files are parsed using the following command.

`perl parse_miRDeep_prediction.pl miRDP2_predictions_list.txt miRDP2`

The file `miRDP2_predictions_list.txt` contains the absolut path to each of the `_predictions` file, e.g. `/path/to/result/sample1/sample1_predictions`. There are two resulted files, named `miRDP2_mature.fa` and `miRDP2_arms.txt`, respectively, with `miRDP2` as the prefix. The sequence IDs in `miRDP2_mature.fa` always begin with the prefix `miRDP2_mature_`, followed by order numbers of the unique miRNA mature sequences. The genomic orgin, precursor and primary miRNA sequence inforamtion can be found in the description field of each sequence entry in `miRDP2_mature.fa`. There are three tab-delineated columns in the `miRDP2_arms.txt` file. The first two columns refer to paris of mature miRNA sequences that are predicted from the same precursors. The third column refer to information of the genomic origin, precursor and primary sequences, in the format of `genomic_origin:precursor_sequence:primary_sequence;`. There may be multiple genomic origins, precursor or primary sequences corresponding to the same pairs of miRNA mature sequences.

## 2. 
