# miRDeep-P2_pipeline
This is a general pipeline for performing miRNA prediction and differential expression analysis using small RNA sequencing (sRNA-seq) data of plant samples. In this pipeline, the tool for miRNA prediction from sRNA-seq data is miRDeep-P2 (http://www.srnaworld.com/miRDP2/index.html). The output of miRDeep-P2 is parsed by parse_miRDeep_prediction.pl script in this pipeline. The parsed predicted miRNA sequences are used for  differential expression analysis. In addition, it is recommended to merge the predicted miRNA sequences with those present in the public domain, such as miRBase, for differential expression analysis, in order to increase the sensitivity. After getting the miRNA sequences, the sRNA-seq read data are mapped using Bowtie (http://bowtie-bio.sourceforge.net/index.shtml) and the read count for each miRNA sequence is generated using the bam2ref_counts.pl script in this pipeline. Finally, with the miRNA count data, DESeq2 (https://bioconductor.org/packages/release/bioc/html/DESeq2.html) is used to perform data filtering, normalization and differential expression analysis. 
